src/persist/src/indexed/mod.rs

```diff
@@ -239,18 +240,39 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
         updates.sort_unstable_by(|((k1, v1), t1, _), ((k2, v2), t2, _)| {
             (t1, k1, v1).cmp(&(t2, k2, v2))
         });
-        let batch = BlobFutureBatch {
-            id,
-            desc: Description::new(
-                Antichain::from_elem(desc.start),
-                Antichain::from_elem(desc.end),
-                // We never compact BlobFuture, so since is always the minimum.
-                Antichain::from_elem(SeqNo(0)),
-            ),
-            updates,
-        };
-        let key = self.new_blob_key();
-        self.append_future(key, batch)?;
+
+        // Future batches are required to be consolidated and only have
+        // one update per (ts, k, v).
+        let mut consolidated_updates = vec![];
+
+        for ((t, k, v), group) in &updates
+            .into_iter()
+            .group_by(|((k, v), t, _)| (*t, k.clone(), v.clone()))
+        {
+            let sum = group
+                .map(|((_, _), _, diff)| diff)
+                .reduce(|a, b| a + b)
+                .unwrap_or(0);
+
+            if sum != 0 {
+                consolidated_updates.push(((k, v), t, sum));
+            }
+        }
+
+        if !consolidated_updates.is_empty() {
```

Comment:
nit: This particular case is not a big difference either way, but in general, I find an early return style makes the flow much easier to follow.

```
if consolidated_updates.is_empty() {
    return Ok(());
}
```

Reply:
happy to do that!